### PR TITLE
Add Redux Toolkit

### PR DESCRIPTION
**Pull request by VovaFouchek**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,19 +1,16 @@
...
 import './App.css';
 
 const App = () => {
-  const [list, setList] = useState([]);
   return (
     <div className="App">
       <Wrapper>
         <Routes>
-          <Route path="/" element={<Board list={list} setList={setList} />} />
+          <Route path="/" element={<Board />} />
```
Comment (LessGrits) [inline src/App.jsx:13]: element -> component

File: src/components/Board/index.jsx [hunk 2]
```diff
@@ -1,78 +1,41 @@
...
-import { Leader } from '../Leader';
-import { orderTypes } from '../../helpers/consts';
-import { Controls } from '../Controls';
-import * as helpersFuncions from '../../helpers/functions';
-import { getLeaders } from '../../shared/api/requests/leaders';
-import { Header } from '../Header';
 
+import { Header } from '../Header';
+import { Controls } from '../Controls';
+import { Leader } from '../Leader';
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: import all from one` index.js`

File: src/components/Board/index.jsx [hunk 3]
```diff
@@ -1,78 +1,41 @@
...
+import { ordinalSuffixOf } from '../../helpers/functions';
+import { fetchLeaders } from '../../redux/actions';
+import { addLeaders, editLeaders } from '../../redux/reducer';
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: add alias  in `package.json`

File: src/components/Board/index.jsx [hunk 4]
```diff
@@ -1,78 +1,41 @@
...
 
   const editLeaderInList = editLeader => {
-    const newList = list.map(leader => {
-      if (leader.id === editLeader.id) {
-        return editLeader;
-      }
-      return leader;
-    });
-    setList(refreshList(newList, sortType));
+    const newLeadersList = leaders.map(leader => (leader.id === editLeader.id ? editLeader : leader));
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: move this in reducer

File: src/components/Board/index.jsx [hunk 5]
```diff
@@ -1,78 +1,41 @@
...
-    const listWithNewLeader = [...list, newList];
-    setList(refreshList(listWithNewLeader, sortType));
-  };
-
-  const sortList = () => {
-    const newsortType = reverseOrderType(sortType);
-    setsortType(newsortType);
-    setList(sortListByOrder(list, newsortType));
+    const newLeader = { ...values, position: leaders.length + 1 };
+    const listWithNewLeader = [...leaders, newLeader];
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: reducer

File: src/components/Board/index.jsx [hunk 6]
```diff
@@ -102,14 +56,18 @@ export const Board = ({ list = [], setList }) => {
                 </td>
               </tr>
             )}
+            {/* {error && <h1>Error...</h1>} */}
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: delete

File: src/components/Board/index.jsx [hunk 7]
```diff
@@ -102,14 +56,18 @@ export const Board = ({ list = [], setList }) => {
...
+            {leaders.map(leader => (
+              <Leader
+                leader={leader}
+                editLeaderInList={editLeaderInList}
+                position={ordinalSuffixOf(leader.position)}
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: duplicate data from `leader` object

File: src/components/Header/index.jsx [hunk 8]
```diff
@@ -1,26 +1,31 @@
...
     <>
       <div className={s.banner}>
         <p className={s.text}>All time Highest Scorers</p>
         <div className={s.container__card}>
-          {leaderTop?.length ? (
-            leaderTop.map(leader => <CardLeaderTop leader={leader} key={leader.id} />)
+          {leadersTop?.length ? (
+            leadersTop.map(leader => <CardLeaderTop leader={leader} key={leader.id} />)
           ) : (
             <Box className={s.wrap__progress}>
```
Comment (LessGrits) [inline src/components/Header/index.jsx:21]: use isLoading from store, if no leader write  `no leader`

File: src/components/LeaderProfile/index.jsx [hunk 9]
```diff
@@ -1,23 +1,22 @@
...
   const routers = pathRouters();
-  const [leaderInfo, setLeaderInfo] = useState('');
+  const [leaderInfo, setLeaderInfo] = useState(leaderDefaultValue);
 
-  const getData = async () => {
-    const leadersList = await getLeaders();
-    const leader = leadersList.find(item => item.id === id);
-    setLeaderInfo(leader);
-  };
+  const getDataLeaderProfile = async () => setLeaderInfo(await getLeaderByID(id));
```
Comment (LessGrits) [inline src/components/LeaderProfile/index.jsx]: ?

File: src/redux/actions.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import axios from 'axios';
+import { setZeroScore } from '../helpers/functions';
+import { axiosConfig } from '../shared/api/config/axios';
+import { API } from '../shared/api/config/leaders.api';
+import { startLoading, initLeaders, leadersFetchingError } from './reducer';
+
+export const fetchLeaders = () => async dispatch => {
+  try {
+    dispatch(startLoading());
+    const response = await axios.get(`${axiosConfig.baseURL}${API.LEADERS}`);
```
Comment (LessGrits) [inline src/redux/actions.js]: investigate how to improve this logic with `baseUrl`

File: src/redux/reducer.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+const initialState = {
+  leaders: [],
+  leadersTop: [],
+  sortType: orderTypes.ascending,
+  isLoading: false,
+  error: '',
+};
+
+const setPositions = (sortedList, sortType = initialState.sortType) => {
+  return sortedList.map((item, index) => {
```
Comment (LessGrits) [inline src/redux/reducer.js]: `(` -> `return `

File: src/redux/reducer.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import { createSlice } from '@reduxjs/toolkit';
+import { orderTypes } from '../helpers/consts';
+import { reverseOrderType, sortListByOrder, getLeaderTop } from '../helpers/functions';
+
+const initialState = {
+  leaders: [],
+  leadersTop: [],
```
Comment (LessGrits) [inline src/redux/reducer.js]: for getting leaderTop  useSelect from leaders

File: src/redux/reducer.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  sortType: orderTypes.ascending,
+  isLoading: false,
+  error: '',
```
Comment (LessGrits) [inline src/redux/reducer.js]: null -> ' '

File: src/redux/reducer.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+const setPositions = (sortedList, sortType = initialState.sortType) => {
+  return sortedList.map((item, index) => {
+    if (sortType === orderTypes.ascending) {
+      return { ...item, position: index + 1 };
+    }
+    return { ...item, position: sortedList.length - index };
+  });
```
Comment (LessGrits) [inline src/redux/reducer.js]: .map((item,index) => ({ ...item, position: sortType === orderTypes.ascending ? index + 1 : sortedList.length - index }));

File: src/redux/reducer.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+const leaderSlice = createSlice({
+  name: 'leader',
+  initialState,
+  reducers: {
+    startLoading(state) {
+      state.isLoading = true;
+    },
+    initLeaders(state, action) {
+      state.leaders = sortListByOrder(action.payload, state.sortType);
+      state.leaders = setPositions(action.payload, state.sortType);
```
Comment (LessGrits) [inline src/redux/reducer.js]: use variable for editing data and in  the end set data only one time

File: src/redux/reducer.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+      state.isLoading = false;
+    },
+    leadersFetchingError(state, action) {
+      state.error = action.payload;
+      state.isLoading = true;
```
Comment (LessGrits) [inline src/redux/reducer.js]: loading = false

File: src/redux/reducer.js [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    sortLeaders(state) {
+      state.sortType = reverseOrderType(state.sortType);
+      state.leaders = sortListByOrder(state.leaders, state.sortType);
+    },
+    addLeaders(state, action) {
+      state.leaders = sortListByOrder(action.payload, state.sortType);
+      state.leaders = setPositions(action.payload, state.sortType);
```
Comment (LessGrits) [inline src/redux/reducer.js]: look above

File: src/redux/reducer.js [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    editLeaders(state, action) {
+      state.leaders = sortListByOrder(action.payload, state.sortType);
+      state.leaders = setPositions(action.payload, state.sortType);
```
Comment (LessGrits) [inline src/redux/reducer.js]: look above

File: src/redux/reducer.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    initLeadersTop(state, action) {
+      state.leadersTop = getLeaderTop(action.payload);
+    },
```
Comment (LessGrits) [inline src/redux/reducer.js]: useSelect

File: src/redux/store.js [hunk 20]
```diff
@@ -0,0 +1,10 @@
+import { configureStore } from '@reduxjs/toolkit';
+import logger from 'redux-logger';
+import leaderReducer from './reducer';
+
+export const store = configureStore({
+  reducer: {
+    leader: leaderReducer,
+  },
+  middleware: getDefaultMiddleware => getDefaultMiddleware({ serializableCheck: false }).concat(logger),
```
Comment (LessGrits) [inline src/redux/store.js]: ?

File: src/shared/api/requests/leaders.js [hunk 21]
```diff
@@ -10,16 +11,24 @@ export const getLeaders = async () => {
   }
 };
 
+export const getLeaderByID = async id => {
+  try {
+    const response = await axios.get(`${axiosConfig.baseURL}${API.LEADERS}/${id}`);
```
Comment (LessGrits) [inline src/shared/api/requests/leaders.js]: look above

File: src/shared/api/requests/leaders.js [hunk 22]
```diff
@@ -10,16 +11,24 @@ export const getLeaders = async () => {
...
+    return response.data;
+  } catch (error) {
+    return null;
```
Comment (LessGrits) [inline src/shared/api/requests/leaders.js]: throw error

File: src/redux/reducer.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+    leadersFetchingError(state, action) {
+      state.error = action.payload;
+      state.isLoading = true;
+    },
+    sortLeaders(state) {
+      state.sortType = reverseOrderType(state.sortType);
+      state.leaders = sortListByOrder(state.leaders, state.sortType);
+    },
+    addLeaders(state, action) {
```
Comment (LessGrits) [inline src/redux/reducer.js]: add leader (one)

File: src/redux/reducer.js [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+      state.leaders = sortListByOrder(action.payload, state.sortType);
+      state.leaders = setPositions(action.payload, state.sortType);
+    },
+    editLeaders(state, action) {
```
Comment (LessGrits) [inline src/redux/reducer.js]: look above

File: src/App.jsx (hunk 1 above)
Comment (VovaFouchek) [reply src/App.jsx:13]: "Component" used in Route v4 and I wrote, that in v6, it's a good case

File: src/components/Board/index.jsx [hunk 25]
```diff
@@ -1,78 +1,31 @@
...
-import { Leader } from '../Leader';
-import { orderTypes } from '../../helpers/consts';
-import { Controls } from '../Controls';
-import * as helpersFuncions from '../../helpers/functions';
-import { getLeaders } from '../../shared/api/requests/leaders';
-import { Header } from '../Header';
 
+import { fetchLeaders } from 'redux/actions';
+import { addLeader, editLeader } from 'redux/reducer';
+import { Header, Controls, Leader } from '../index';
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: alias components

File: src/components/Board/index.jsx [hunk 26]
```diff
@@ -1,78 +1,31 @@
...
+  const editLeaderInList = leader => dispatch(editLeader(leader, sortType));
+
+  const addLeaderInList = leader => dispatch(addLeader(leader, sortType));
+
   return (
     <>
-      <Header list={list} setList={setList} />
+      <Header />
       <div className={s.board}>
-        <Controls sortListByOrder={sortList} addLeaderInList={addLeaderInList} />
```
Comment (LessGrits) [inline src/components/Board/index.jsx:75]: use dispatch inside component

File: src/components/Board/index.jsx [hunk 27]
```diff
@@ -1,78 +1,31 @@
...
+        <Controls addLeaderInList={addLeaderInList} />
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: see above

File: src/components/Board/index.jsx [hunk 28]
```diff
@@ -102,14 +46,14 @@ export const Board = ({ list = [], setList }) => {
                 </td>
               </tr>
             )}
+            {leaders.map(leader => (
+              <Leader leader={leader} editLeaderInList={editLeaderInList} key={leader.id} />
```
Comment (LessGrits) [inline src/components/Board/index.jsx]: see above

Comment (LessGrits) [reply src/components/Board/index.jsx]: mb  ...inList is excessive

File: src/components/Controls/index.jsx [hunk 29]
```diff
@@ -1,17 +1,24 @@
...
 import s from './controls.module.scss';
 
-export const Controls = ({ sortListByOrder, addLeaderInList }) => {
+const Controls = ({ addLeaderInList }) => {
+  const dispatch = useDispatch();
   const [open, setOpen] = useState(false);
+
   const handleOpen = () => setOpen(true);
   const handleClose = () => setOpen(false);
+  const dispatchSortLeaders = () => dispatch(sortLeaders());
```
Comment (LessGrits) [inline src/components/Controls/index.jsx]: not use dispatch inside variable name

File: src/components/Controls/index.jsx [hunk 30]
```diff
@@ -1,17 +1,24 @@
...
-import { BasicModal } from '../../shared/components/BasicModal';
+import { useDispatch } from 'react-redux';
+
+import { BasicModal } from 'shared/components/BasicModal';
+import { sortLeaders } from 'redux/reducer';
+import AddForm from '../AddForm';
 import s from './controls.module.scss';
 
-export const Controls = ({ sortListByOrder, addLeaderInList }) => {
+const Controls = ({ addLeaderInList }) => {
```
Comment (LessGrits) [inline src/components/Controls/index.jsx]: get it from useDispatch

File: src/components/Controls/index.jsx [hunk 31]
```diff
@@ -25,6 +32,7 @@ export const Controls = ({ sortListByOrder, addLeaderInList }) => {
 };
 
 Controls.propTypes = {
-  sortListByOrder: PropTypes.func.isRequired,
   addLeaderInList: PropTypes.func.isRequired,
```
Comment (LessGrits) [inline src/components/Controls/index.jsx]: remove

File: src/components/Leader/index.jsx [hunk 32]
```diff
@@ -55,6 +55,7 @@ Leader.propTypes = {
   leader: PropTypes.shape({
     name: PropTypes.string.isRequired,
```
Comment (LessGrits) [inline src/components/Leader/index.jsx:56]: is leader has only name?

**Pull request merged**